### PR TITLE
Add a method to get estimated export size in IMigrator

### PR DESCRIPTION
**Pull request by come-nc**
Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

### CODE REVIEW

**Comment by blizzz** [comment]
maser is 25 :)

File: apps/dav/lib/UserMigration/ContactsMigrator.php [hunk 1]
```diff
@@ -193,6 +193,21 @@ private function serializeCards(array $vCards): string {
...
+	/**
+	 * {@inheritDoc}
+	 */
+	public function getExportEstimatedSize(IUser $user): int {
+		$principalUri = $this->getPrincipalUri($user);
+
+		return array_sum(array_map(
+			function (array $addressBookInfo) use ($user): int {
+				// FIXME 1MiB by addressbook, no idea if this is accurate and if we should go into more details
+				return 1000;
```
Comment (tcitworld) [inline apps/dav/lib/UserMigration/ContactsMigrator.php:205]: Possibly more because contacts can hold base64-encoded pictures.

Comment (come-nc) [reply apps/dav/lib/UserMigration/ContactsMigrator.php:205]: Can you think of some good way of estimating the size of exported data without taking too much time?

Comment (skjnldsv) [reply apps/dav/lib/UserMigration/ContactsMigrator.php:205]: Seconded, this is complicated to estimate.

Comment (Pytal) [reply apps/dav/lib/UserMigration/ContactsMigrator.php:205]: @tcitworld same for calendar as well? number of components?

**Comment by Pytal** [comment]
Using a sample calendar that would be representative a regular user's calendar and querying
```sql
select avg(size) from oc_calendarobjects where calendarid = <id>
```
resulted in `~450B`, while using a sample address book without images and querying
```sql
select avg(size) from oc_cards where addressbookid = <id>
```
resulted in `~350B`

Base64 encoded contact image sizes can't be quickly estimated for now given that uploaded images are not re-encoded on upload to target a size range, so estimation of those are not taken into account

File: apps/dav/lib/UserMigration/CalendarMigrator.php [hunk 2]
```diff
@@ -206,6 +207,31 @@ private function getUniqueCalendarUri(IUser $user, string $initialCalendarUri):
 		return $calendarUri;
 	}
 
+	/**
+	 * {@inheritDoc}
+	 */
+	public function getEstimatedExportSize(IUser $user): int {
+		$calendarExports = $this->getCalendarExports($user, new NullOutput());
```
Comment (tcitworld) [inline apps/dav/lib/UserMigration/CalendarMigrator.php:214]: This goes through a lot just to get a count.
Wouldn't an new `countCalendarObjectsForUser` method in `CalDAVBackend` do the job instead?

Comment (Pytal) [reply apps/dav/lib/UserMigration/CalendarMigrator.php:214]: The execution is still fast in under a second for the entire export which goes through the same code paths and the returned export size will be cached higher up with https://github.com/nextcloud/user_migration/pull/135/commits/fcdcb6b2deda99ddf4414ad1980f7159e5700ab5 to make it snappy on subsequent calls, so let's keep this PR in scope unless you could make the additions to `CalDAVBackend` as you're more familiar?

Comment (Pytal) [reply apps/dav/lib/UserMigration/CalendarMigrator.php:214]: The execution is still fast in under a second for the entire export which goes through the same code paths and the returned export size will be cached higher up with https://github.com/nextcloud/user_migration/pull/135/commits/fcdcb6b2deda99ddf4414ad1980f7159e5700ab5 to make it snappy on subsequent calls, so let's not scope creep this PR unless you could make the additions to `CalDAVBackend` as you're more familiar?

**Comment by come-nc** [comment]
- [ ] Move new method to new interface (capability mode)

(This is because 24 is now out we cannot modify IMigrator)

**Comment by come-nc** [comment]
@Pytal Ready for review

**Comment by Pytal** [comment]
Commits overwritten by force push have been restored

**Review by Pytal — APPROVED** [review]

**Review by CarlSchwan — APPROVED** [review]

**Comment by Pytal** [comment]
Updated autoloaders

**Comment by Pytal** [comment]
CI failure unrelated

**Pull request merged**